### PR TITLE
Compiler now catches when an expression is raised which is no Exception

### LINKED ISSUE #6805: Raising object not inherited from Exception should not compile
Example code:
https://play.nim-lang.org/?gist=f9efb940bd32980a30d86992eb2a993e

When raising the object not inherited from Exception, program crashes.
It seems to me that when compiler can detect that, it would be useful that to result in compilation error.

It's easy to forget to write that "of Exception", and as exceptions are rare events which may be not covered by tests, it's easy to keep that mistype unnoticed.

### PR DESCRIPTION
**Pull request by Veladus**
Fixes #6805.

### CODE REVIEW

File: compiler/semstmts.nim [hunk 1]
```diff
@@ -729,6 +729,16 @@ proc semRaise(c: PContext, n: PNode): PNode =
     var typ = n.sons[0].typ
     if typ.kind != tyRef or typ.lastSon.kind != tyObject:
       localError(n.info, errExprCannotBeRaised)
+    
+    # check if the given object inherits from Exception
+    var base = typ.lastSon
+    while true:
+      if base.sym.name.s == "Exception":
```
Comment (Araq) [inline compiler/semstmts.nim]: That is not enough, you need to at least also check ``base.owner.flags.contains(sfSystemModule)``. But the clean solution is to mark ``Exception`` with a new magic that you check for here.

File: compiler/msgs.nim [hunk 2]
```diff
@@ -269,6 +269,7 @@ const
     errExprHasNoAddress: "expression has no address",
     errXStackEscape: "address of '$1' may not escape its stack frame",
     errVarForOutParamNeededX: "for a \'var\' type a variable needs to be passed; but '$1' is immutable",
+    errExprIsNoException: "raised object does not inherit from Exception",
```
Comment (Araq) [inline compiler/msgs.nim]: We don't do this anymore. Use the version of ``localError`` that takes a string.

File: compiler/semstmts.nim [hunk 3]
```diff
@@ -729,6 +729,16 @@ proc semRaise(c: PContext, n: PNode): PNode =
...
+        break
+      if base.lastSon == nil:
+        localError(n.info, errExprIsNoException)
```
Comment (Araq) [inline compiler/semstmts.nim]: Good error message: Write out which object type does not inherit from exception.

File: compiler/semstmts.nim (hunk 1 above)
Comment (Veladus) [reply compiler/semstmts.nim]: Okay, is there any good documentation of how to use magic in the compiler or could you point me to a file, where this is used extensively?

Comment (Araq) [reply compiler/semstmts.nim]: add 'mException' to the TMagic enum in ast.nim and check for it here. Mark Exception as ``magic: Exception``

Comment (Veladus) [reply compiler/semstmts.nim]: Tried this, but I can get the magic mException to work on Exception. I added it to TMagic and can use it in code (tried echo), but when I write `{.magic: "Exception", compilerproc.}` or `{.magic Exception, compilerproc.}` i get *Error: type expected* immediatly on this line.

Comment (Veladus) [reply compiler/semstmts.nim]: Tried this, but I can't get the magic mException to work on Exception. I added it to TMagic and can use it in code (tried echo), but when I write `{.magic: "Exception", compilerproc.}` or `{.magic Exception, compilerproc.}` on Exception, I get *Error: type expected* immediatly on this line.

Comment (Araq) [reply compiler/semstmts.nim]: For me 
```nim

  Exception* {.compilerproc, magic: "Exception".} = object of RootObj
```
works.

**Comment by cooldome** [comment]
Looks neat, I think it is also required to support throw/catch for native cpp/javascript exceptions not only Nim exceptions

**Comment by Veladus** [comment]
@cooldome I think this should be checked too, but it's not topic of this PR and I'm certainly not the right person to tackle this.
Atleast for cpp exceptions there is an open issue concerning them #3571

File: compiler/semstmts.nim (hunk 1 above)
Comment (Veladus) [reply compiler/semstmts.nim]: But it doesn't work for me :/

I've added `mException` to TMagic in ast.nim.
Than I ran `.\koch boot` which compiled fine and in this state I can use `TMagic.mException`.
Than I copy pasted your line and run `.\koch boot` and I get the aformentioned error.

The error is 

    > lib\system.nim(466, 3) Error: type expected

so the error doesn't complain about the pragma directly, but rather about `Exception` as a type, but weirdly when I remove the `magic` pragma the message disapears

File: lib/system.nim [hunk 4]
```diff
@@ -463,7 +463,7 @@ type
     line*: int          ## line number of the proc that is currently executing
     filename*: cstring  ## filename of the proc that is currently executing
 
-  Exception* {.compilerproc.} = object of RootObj ## \
+  Exception* {.compilerproc, magic: "Exception".} = object of RootObj ## \
```
Comment (Veladus) [inline lib/system.nim]: @Araq what do I have to change to make this work?

**Pull request merged**